packages/tools/client-debugger/src/FluidClientDebugger.ts

```diff
@@ -0,0 +1,354 @@
+    public getMinimumSequenceNumber(): number {
+        return this.container.deltaManager.minimumSequenceNumber;
+    }
+
+    public getOpsLog(): readonly ISequencedDocumentMessage[] {
+        // Clone array contents so consumers don't see local changes
+        return this._opsLog.map((value) => value);
+    }
+
+    private onIncomingOpProcessed(op: ISequencedDocumentMessage): void {
+        this._opsLog.push(op);
+
+        this.incomingOpProcessedHandler(op);
+    }
+
+    // #endregion
+
+    // #region Audience data
+
+    /**
+     * {@inheritDoc IFluidClientDebugger.getAudienceMembers}
+     */
+    public getAudienceMembers(): Map<string, IClient> {
+        return this.audience.getMembers();
+    }
+
+    /**
+     * {@inheritDoc IFluidClientDebugger.getAuidienceHistory}
+     */
+    public getAudienceHistory(): readonly AudienceChangeLogEntry[] {
+        // Clone array contents so consumers don't see local changes
+        return this._audienceChangeLog.map((value) => value);
+    }
+
+    private onAudienceMemberAdded(clientId: string, client: IClient): void {
+        this._audienceChangeLog.push({
+            clientId,
+            client,
+            changeKind: "added",
+            timestamp: Date.now(),
+        });
+        this.audienceMemberChangeHandler(MemberChangeKind.Added, clientId, client);
+    }
+
+    private onAudienceMemberRemoved(clientId: string, client: IClient): void {
+        this._audienceChangeLog.push({
+            clientId,
+            client,
+            changeKind: "removed",
+            timestamp: Date.now(),
+        });
+        this.audienceMemberChangeHandler(MemberChangeKind.Removed, clientId, client);
+    }
+
+    // #endregion
+
+    // #region User actions
+
+    /**
+     * {@inheritDoc IFluidClientDebugger.disconnectContainer}
+     */
+    public disconnectContainer(): void {
+        // TODO: Provide along reason string once API is updated to accept one.
+        this.container.disconnect();
+    }
+
+    /**
+     * {@inheritDoc IFluidClientDebugger.tryConnectContainer}
+     */
+    public tryConnectContainer(): void {
+        this.container.connect();
+    }
+
+    /**
+     * {@inheritDoc IFluidClientDebugger.closeContainer}
+     */
+    public closeContainer(): void {
+        // TODO: Provide reason string if/when the close API is updated to accept non-error "reason"s.
+        this.container.close();
+    }
+
+    // #endregion
+
+    /**
+     * {@inheritDoc IFluidClientDebugger.dispose}
+     */
+    public dispose(): void {
+        // Bind Container events
+        this.container.off("attached", () => this.onContainerAttached());
+        this.container.off("connected", (clientId) => this.onContainerConnected(clientId));
+        this.container.off("disconnected", () => this.onContainerDisconnected());
+        this.container.off("closed", (error) => this.onContainerClosed(error));
+        this.container.off("op", (op) => this.onIncomingOpProcessed(op));
```

Comment:
none of these will actually remove the handler as you are creating a new function. you need turn off the same instance of the function you are registering

Reply:
Ah, that's right. I'll fix this. Thanks!

Reply:
Fixed on the dev branch